### PR TITLE
Fix matching regex word-boundary (\b) in strings replace

### LINKED ISSUE #9950: [FEA] Support replacing word boundaries in regexp replace in way that is compatible with Python/Java
**Is your feature request related to a problem? Please describe.**
When replacing the pattern `\\b` (word boundary) with `X` for the string `a\nb`, Python and Java produce `XaX\nXbX` and cuDF produces `XXXa\nb`. 

```python
>>> re.sub('\\b', 'X', 'a\nb')
'XaX\nXbX'
```

```python
>>> cudf.Series(['a\nb']).str.replace('\\b', 'X', regex=True)
0    XXXa\nb
```

**Describe the solution you'd like**
I would like the ability to match Python/Java behavior in this case.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #9950 

Fixes matching a single word-boundary (BOW) regex pattern. This pattern will match word boundaries and not any actual characters. This means the `(begin,end)` position values will be equal. The replace code was always expecting `begin < end` character range to replace. The logic has been updated to allow for this case.

Additional gtests have been added that include a single `\b` pattern character.

### CODE REVIEW

File: cpp/src/strings/replace/replace_re.cu [hunk 1]
```diff
@@ -62,37 +62,44 @@ struct replace_regex_fn {
...
-    auto in_ptr       = d_str.data();                    // input pointer (i)
-    auto out_ptr      = d_chars ? d_chars + d_offsets[idx] : nullptr;  // output pointer (o)
-    size_type lpos    = 0;
-    int32_t begin     = 0;
-    int32_t end       = static_cast<int32_t>(nchars);
+
+    auto const d_str = d_strings.element<string_view>(idx);
+    auto nbytes      = d_str.size_bytes();                  // number of bytes in input string
+    auto mxn = maxrepl < 0 ? d_str.length() + 1 : maxrepl;  // max possible replaces for this string
+    auto in_ptr  = d_str.data();                            // input pointer (i)
```
Comment (bdice) [inline cpp/src/strings/replace/replace_re.cu:69]: Can some more of these values be `const`?

**Review by bdice — APPROVED** [review]
LGTM - some minor improvements to `const`ness might be possible, but those are non-blocking.

File: cpp/src/strings/replace/replace_re.cu (hunk 1 above)
Comment (davidwendt) [reply cpp/src/strings/replace/replace_re.cu:69]: I believe everything that can be const is declared const. I will double-check.
The names here actually help with the code formatting so I can line up comments without clang-format making a mess of my code.

Comment (bdice) [reply cpp/src/strings/replace/replace_re.cu:69]: I just re-checked a couple of things that I thought could be made `const` but they cannot. This is fine as-is!

**Comment by davidwendt** [comment]
rerun tests

File: cpp/src/strings/replace/replace_re.cu [hunk 2]
```diff
@@ -62,37 +62,49 @@ struct replace_regex_fn {
...
-      end   = static_cast<int32_t>(nchars);
+      if (prog.is_empty() || prog.find<stack_size>(idx, d_str, begin, end) <= 0) {
+        break;  // no more matches
+      }
+
+      auto const start_pos = d_str.byte_offset(begin);        // get offset for these
+      auto const end_pos   = d_str.byte_offset(end);          // character position values
+      nbytes += d_repl.size_bytes() - (end_pos - start_pos);  // and compute new size
+
+      if (out_ptr)                                           // replace:
```
Comment (ttnghia) [inline cpp/src/strings/replace/replace_re.cu:87]: Style issue....

File: cpp/src/strings/replace/replace_re.cu [hunk 3]
```diff
@@ -62,37 +62,49 @@ struct replace_regex_fn {
...
+    auto in_ptr        = d_str.data();                      // input pointer (i)
+    auto out_ptr       = d_chars ? d_chars + d_offsets[idx]  // output pointer (o)
+                                 : nullptr;
+    size_type last_pos = 0;
+    int32_t begin      = 0;   // these are for calling prog.find
+    int32_t end        = -1;  // matches final word-boundary if at the end of the string
+
     // copy input to output replacing strings as we go
     while (mxn-- > 0)  // maximum number of replaces
     {
```
Comment (ttnghia) [inline cpp/src/strings/replace/replace_re.cu:78]: I believe that this is also a style issue (curly bracket needs to be at the same line of `if/while...`).

File: cpp/src/strings/replace/replace_re.cu [hunk 4]
```diff
@@ -62,37 +62,49 @@ struct replace_regex_fn {
...
+      }                                                      //  in_ptr --^
+
+      begin = end + (begin == end);
+      end   = -1;
     }
-    if (out_ptr)                                                  // copy the remainder
-      memcpy(out_ptr, in_ptr + lpos, d_str.size_bytes() - lpos);  // o:bbbbrrrrrreeee
-    else
+
+    if (out_ptr) {
```
Comment (mythrocks) [inline cpp/src/strings/replace/replace_re.cu:101]: Thank you for this formatting change. I know you'd prefer not to use braces here.

File: cpp/src/strings/replace/replace_re.cu [hunk 5]
```diff
@@ -62,37 +62,49 @@ struct replace_regex_fn {
...
-        lpos = epos;                                                        // i:bbbbsssseeee
-      }                                                                     //  in_ptr --^
-      begin = end;
-      end   = static_cast<int32_t>(nchars);
+      if (prog.is_empty() || prog.find<stack_size>(idx, d_str, begin, end) <= 0) {
+        break;  // no more matches
+      }
+
+      auto const start_pos = d_str.byte_offset(begin);        // get offset for these
+      auto const end_pos   = d_str.byte_offset(end);          // character position values
```
Comment (ttnghia) [inline cpp/src/strings/replace/replace_re.cu:84]: I'm not sure if this is still valid when `end = -1` (set above)?

**Review by mythrocks — APPROVED** [review]
LGTM.

**Comment by galipremsagar** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**